Core/Core/SwiftUIViews/Pages.swift

```diff
@@ -0,0 +1,96 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2020-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+
+@available(iOSApplicationExtension 13.0, *)
+public struct Pages<Item: Identifiable, Content: View>: View {
+    public typealias ContentBuilder = (Item) -> Content
+
+    let items: [Item]
+    let content: ContentBuilder
+    @Binding var currentIndex: Int
+
+    @Environment(\.layoutDirection) var layoutDirection
+    @GestureState var translation: CGFloat = 0
+
+    var spacing: CGFloat = 0
+    var scaling: (CGFloat) -> CGFloat = { _ in 1 }
+
+    var minIndex: Int { max(0, currentIndex - 1) }
+    var maxIndex: Int { min(items.count, currentIndex + 2) }
+    var dx: CGFloat { layoutDirection == .rightToLeft ? -1 : 1 }
+
+    func show(index: Int) {
+        currentIndex = min(max(index, 0), items.count - 1)
+    }
+
+    func scale(for item: Item, width: CGFloat) -> CGFloat {
+        guard let index = items.firstIndex(where: { $0.id == item.id }) else { return 1 }
```

Comment:
Might be worth also guarding on `width` being greater than 0 to avoid divide by 0. In case this is ever called with a computed property like the width of a view that ends up being 0 during some state change or something.

Reply:
Good thing to check. I'll add that.